### ISSUE TITLE
IDA 6.8: The newprc priority hook for initializing the typemapper doesn't seem to work

### ISSUE THREAD

**Issue by gool123456** [arizvisa/ida-minsc#46]
This plugin is really cool !! But he does not support ida6.8, 

Ida6.8 required for company work !  Can it be resolved and let it support ? thk 

that installed show this error log :

```

internal.interface.priorityhook.apply('newprc') : Unable to apply the hook for an unknown method.
Traceback (most recent call last):
  File "D:\tools\IDA_Pro_v6.8_and_Hex-Rays_Decompiler_(ARM,x64,x86)_Hskyre\python\idaapi.py", line 601, in IDAPython_ExecScript
    execfile(script, g)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\idapythonrc.py", line 198, in <module>
    __root__ = imp.load_source('__root__', os.path.join(root, '__root__.py'))
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\__root__.py", line 108, in <module>
    ui.hook.idp.add('newprc', __import__('internal').interface.typemap.__newprc__, 0)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\_interface.py", line 298, in add
    res = self.apply(name)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\_interface.py", line 338, in apply
    raise NameError("{:s}.apply({!r}) : Unable to apply the hook for an unknown method.".format('.'.join(('internal', __name__, cls.__name__)), name))
NameError: internal.interface.priorityhook.apply('newprc') : Unable to apply the hook for an unknown method.
```

**Comment by arizvisa** [comment]
Awesome, I'm glad you like it! Sorry that I'm getting to this 2 days late, I've been in the process of relocating and it's been beating me up. Hah.

Let me see what I can do. I worked on this library since pre-6.95 and intend for it to remain compatible with older versions of IDA as well, so **yes** if you have any issues w/ older versions of IDA I **will** support it. So definitely let me know if you have an issue or something doesn't work as intended.

Give me a few days (maybe today) and I'll see if I can come up w/ a workaround for this particular issue. I've currently been porting my things (like ida-minsc and such) to deal with the whole Python3 fiasco and so I've been having to deal w/ that distraction.

**Comment by arizvisa** [comment]
(Renaming the title of this issue to represent the symptom)

**Comment by arizvisa** [comment]
@gool123456, I'll be working on your issue in PR #47.

Just ftr, as IDAPython on IDA 6.8 doesn't expose all of the hook typs (and IDA doesn't either). Keep in mind that some of the "magic" that `ida-minsc` is doing won't be 100% possible to get working. I'll add documentation at some point once I finish the PR, but for example:

1. Whenever you create a comment with IDA's UI (using `;` or `:`), ida-minsc updates an internal index so that you can query things w/ `db.select`, `db.selectcontents`, or `func.select`. As we're unable to hook the UI, if you want to leverage the tagging component you'll need to explicitly execute `db.tag('mycomment', 'mywhatever')` or `func.tag('myfunccomment', 42)`.

2. The processor module might not be auto-detectable, so some of the magic inside the `instruction` module for doing register searches might not work either. Things like `ins.op(ea, 1)`, `db.a.prevreg(ea, 'rcx', write=1)`, or `for ea in func.register('rax', write=1): ...`.

In the meantime, can you comment out those hooks which are in `__root__.py` and make sure there aren't any other issues? As I'm fixing this blind (due to not being able to find a copy of IDA 6.8 to dev against), I'm having to cross reference the hooks against the documentation I can find.

**Comment by arizvisa** [comment]
@gool123456, when you get a chance can you try out PR #47 and see if that works for you in its current state?

I ended up implementing a priority queue for `idaapi.notify_when` and use that to try and catch when IDA opens/closes databases. Hopefully some of the older code from pre-6.95 still works...

**Comment by gool123456** [comment]
@arizvisa hi, arizvisa, Nice to see your busy reply :) 

I have also been studying this issue for the past two days. 
This is really a great project. I have learned a lot from it.

I tried to remove the hook in __root__.py, ida6.8 is working. 
But new problems have also occurred. As you said, idapython did not update action. Because we removed the hook.

I loaded the .so file, which is an ARM instruction. Using instruction.op(ea, 0) it gets X86 registers. It doesn't seem to know that I am reverse engineering in ARM

from:

def op(ea, opnum):
    '''Decodes the operand `opnum` for the instruction at the address `ea`.'''
    res = operand(ea, opnum)
    return __optype__.decode(ea, res)


__optype__ has not been updated?


I'm using ida 6.8 from here, and you use qq account to download :

link：https://share.weiyun.com/55M0Rse  access code：fm6d9p

thk!

**Comment by gool123456** [comment]
@arizvisa I'm trying the method you said, I hope to give you good news

**Comment by gool123456** [comment]
@arizvisa 6.8 issue : A NEG dialog is frequently displayed when using certain functions, such as when using database.functions.list ().

**Comment by arizvisa** [comment]
Awesome. Thank you. I'll download it and see what other issues I can find.

Did you try out PR #47 and see if it solved all of the hook problems in __root__.py?

**Comment by gool123456** [comment]
@arizvisa Sorry, I did n’t understand 47 before, and I have been busy with work since then. Today I deliberately tested # 47. He shows that there are still errors: 

`WARNING:root:internal.interface.prioritynotification.add(0x1, <bound method classobj.__nw_newprc__ of <class interface.typemap at 0x06560BC8>>, priority=10) : Unable to connect to the specified notification (0x1).
internal.interface.priorityhook.apply('tail_owner_changed') : Unable to apply the hook for an unknown method.
Traceback (most recent call last):
  File "D:\tools\IDA_Pro_v6.8_and_Hex-Rays_Decompiler_(ARM,x64,x86)_Hskyre\python\idaapi.py", line 601, in IDAPython_ExecScript
    execfile(script, g)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\idapythonrc.py", line 198, in <module>
    __root__ = imp.load_source('__root__', os.path.join(root, '__root__.py'))
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\__root__.py", line 208, in <module>
    ui.hook.idp.add('tail_owner_changed', __import__('hooks').tail_owner_changed, 40)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\_interface.py", line 304, in add
    res = self.apply(name)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\_interface.py", line 350, in apply
    raise NameError("{:s}.apply({!r}) : Unable to apply the hook for an unknown method.".format('.'.join(('internal', __name__, cls.__name__)), name))
NameError: internal.interface.priorityhook.apply('tail_owner_changed') : Unable to apply the hook for an unknown method.`

**Comment by arizvisa** [comment]
Ah ok. That one should be `idb.tail_owner_changed`. Do you mind syncing and trying that PR again? And if you still have issues, can you do the following and add a comment to PR #47.

```
print dir(idaapi.IDB_Hooks)
print dir(idaapi.IDP_Hooks)
print dir(idaapi.UI_Hooks)
```

Those 3 lines of python should emit 3 lists of all of the available events. Essentially what I'm doing is assigning workarounds using the hooks that you have available.

I unfortunately was not able to download the version of IDA that you had linked me to, so I'm having to fix these issues somewhat blind.

**Comment by gool123456** [comment]
Hey hello, I try again, it happened the same.

Then executed the python command:

```
Python> print dir (idaapi.IDB_Hooks)
['__class__', '__del__', '__delattr__', '__dict__', '__disown__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', ' __reduce__ ',' __reduce_ex__ ',' __repr__ ',' __setattr__ ',' __sizeof__ ',' __str__ ',' __subclasshook__ ',' __swig_destroy__ ',' area_um_cmt'cm_cmt , 'enum_cmt_changed', 'enum_created', 'enum_deleted', 'enum_member_created', 'enum_member_deleted', 'enum_renamed', 'func_noret_changed', 'func_tail_appended', 'func_tail_removed', '_op _,' op segm_added ',' segm_deleted ',' segm_end_changed ',' segm_moved ',' segm_start_changed ',' struc_cmt_changed ',' struc_created ',' struc_deleted ',' struc_expanded ',' struc_member_de __'_''_ ' , 'struc_renamed', 'tail_owner_changed', 'thisown', 'thunk_func_created', 'ti_changed', 'unhook']
Python> print dir (idaapi.IDP_Hooks)
['__class__', '__del__', '__delattr__', '__dict__', '__disown__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', ' __reduce__ ',' __reduce_ex__ ',' __repr__ ',' __setattr__ ',' __sizeof__ ',' __str__ ',' __subclasshook__ ',' __swig_destroy__ ',' __weakref__ ',' add_y''auto ', , 'auto_queue_empty', 'closebase', 'custom_ana', 'custom_emu', 'custom_mnem', 'custom_out', 'custom_outop', 'del_func', 'hook', 'is_call_insn', 'is_ret_insn', 'is_sane_insn', ' load_idasgn ',' make_code ',' make_data ',' may_be_func ',' rename ',' renamed ',' savebase ',' thisown ',' undefine ',' unhook ']
Python> print dir (idaapi.UI_Hooks)
['__class__', '__del__', '__delattr__', '__dict__', '__disown__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', ' __reduce__ ',' __reduce_ex__ ',' __repr__ ',' __setattr__ ',' __sizeof__ ',' __str__ ',' __subclasshook__ ',' __swig_destroy__ ',' current_t__form_fin_t_form_find , 'populating_tform_popup', 'postprocess', 'preprocess', 'saved', 'saving', 'term', 'thisown', 'unhook', 'updated_actions', 'updating_actions']
```

I hope to help you download, so that the official can truly support ida6.8, I have uploaded it to Google Dirve, please try again :

https://drive.google.com/open?id=1BJf7hpQZI6x4D2s_Ep4Zg2eyyI7VQnAZ

**Comment by arizvisa** [comment]
Are you sure you synchronized properly and did a git pull?

The only 2 warnings that should be listed is a warning about the NW_INITIDA and NW_TERMIDA commands not being able to be set, the `tail_owner_changed` warnings should be gone as they were due to a type-o. It should give out a warning mentioning completely different files. None of the hooks are in `__root__.py` anymore, so you should see a different file emitting those warnings.

Try typing in `git pull origin GH-46` to synchronize. Once you confirm that it warns about being unable to set notifications on `idaapi.NW_INITIDA` and `idaapi.NW_TERMIDA`, I'll push the final fix which disables those debugging messages.

**Comment by arizvisa** [comment]
Actually, apparently there's a better way to sync from a specific pull request. Just make sure you do a `git fetch origin pull/47/head:GH-47`, then checkout the branch via `git checkout GH-47`. Then to synchronize, `git pull origin pull/47/head`.

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally

**Comment by arizvisa** [comment]
I'm gonna just upload the other two commits. They should emit warnings because for some reason the NW_INITIDA and NW_TERMIDA notifications don't work reliably on some versions of IDA, but the other hooks should still be assigned.

Once you synchronize you should see those two warnings as all the other debugging logs are non-critical. Let me know once you see them.

**Comment by gool123456** [comment]
@arizvisa  Please see;
https://github.com/arizvisa/ida-minsc/pull/47#issuecomment-604393635

**Comment by arizvisa** [comment]
As this issue was fixed by PR #47, do you mind closing this as being resolved?

**Comment by arizvisa** [comment]
Correction. This issue was closed by PR #48 

I'll close this. I don't think any of my users really care anyways. ;)

**Comment by gool123456** [comment]
> As this issue was fixed by PR #47, do you mind closing this as being resolved?

Don't mind, the problem has been solved :)
